File: src/todoist_client.py
"""Todoist API client wrapper."""
# mypy: disable-error-code=attr-defined

import logging
from datetime import datetime
from typing import Any

from todoist_api_python.api import TodoistAPI

logger = logging.getLogger(__name__)


class TodoistClient:
    """Client for interacting with Todoist API."""

    def __init__(self, api_token: str) -> None:
        self.api = TodoistAPI(api_token)

    def get_projects(self) -> list[dict[str, Any]]:
        """Get all projects."""
        try:
            projects = list(self.api.get_projects())
            return [
                {
                    "id": p.id,
                    "name": p.name,
                    "color": p.color,
                    "is_favorite": p.is_favorite,
                }
                for p in projects
            ]
        except Exception as e:
            logger.error("Error getting projects: %s", e)
            return []

    def get_tasks(self, filter: str | None = None) -> list[dict[str, Any]]:
        """Get tasks with optional filter."""
        try:
            if filter:
                # Use Todoist filter syntax
                tasks = list(self.api.get_tasks(filter=filter))
            else:
                tasks = list(self.api.get_tasks())

            return [
                {
                    "id": t.id,
                    "content": t.content,
                    "description": t.description,
                    "project_id": t.project_id,
                    "section_id": t.section_id,
                    "labels": t.labels,
                    "priority": t.priority,
                    "due": {"string": t.due.string} if t.due else None,
                    "is_completed": t.is_completed,
                }
                for t in tasks
            ]
        except Exception as e:
            logger.error("Error getting tasks: %s", e)
            return []

    def create_task(
        self,
        content: str,
        description: str | None = None,
        project_id: str | None = None,
        labels: list[str] | None = None,
        priority: int = 1,
        due_string: str | None = None,
    ) -> dict[str, Any]:
        """Create a new task."""
        try:
            task_data: dict[str, Any] = {"content": content, "priority": priority}

            if description:
                task_data["description"] = description
            if project_id:
                task_data["project_id"] = project_id
            if labels:
                task_data["labels"] = labels
            if due_string:
                task_data["due_string"] = due_string

            task = self.api.add_task(**task_data)
            return {
                "id": task.id,
                "content": task.content,
                "description": task.description,
                "project_id": task.project_id,
                "section_id": task.section_id,
                "labels": task.labels,
                "priority": task.priority,
                "due": {"string": task.due.string} if task.due else None,
                "is_completed": task.is_completed,
            }
        except Exception as e:
            logger.error("Error creating task: %s", e)
            raise

    def update_task(
        self,
        task_id: str,
        content: str | None = None,
        description: str | None = None,
        labels: list[str] | None = None,
        priority: int | None = None,
        due_string: str | None = None,
    ) -> bool:
        """Update a task by ID."""
        try:
            update_data: dict[str, Any] = {}
            if content is not None:
                update_data["content"] = content
            if description is not None:
                update_data["description"] = description
            if labels is not None:
                update_data["labels"] = labels
            if priority is not None:
                update_data["priority"] = priority
            if due_string is not None:
                update_data["due_string"] = due_string

            if update_data:
                self.api.update_task(task_id, **update_data)
                return True
            return False
        except Exception as e:
            logger.error("Error updating task: %s", e)
            return False

    def complete_task(self, task_id: str) -> bool:
        """Complete a task by ID."""
        try:
            self.api.complete_task(task_id)
            return True
        except Exception as e:
            logger.error("Error completing task: %s", e)
            return False

    def uncomplete_task(self, task_id: str) -> bool:
        """Reopen a completed task by ID."""
        try:
            self.api.reopen_task(task_id)
            return True
        except Exception as e:
            logger.error("Error uncompleting task: %s", e)
            return False

File: src/todoist-client.ts
import { TodoistApi } from '@doist/todoist-api-typescript';
import type { 
  Task, 
  Project, 
  Section, 
  Label, 
  Comment 
} from '@doist/todoist-api-typescript';

export class TodoistClient {
  private api: TodoistApi;

  constructor(apiToken: string) {
    this.api = new TodoistApi(apiToken);
  }

  // Task methods
  async getTasks(filter?: string): Promise<Task[]> {
    if (filter) {
      // Use the filter endpoint for queries
      const response = await this.api.getTasksByFilter({
        query: filter,
        limit: 200
      });
      return response.results;
    } else {
      // Get all tasks
      const response = await this.api.getTasks({
        limit: 200
      });
      return response.results;
    }
  }

  async getTask(taskId: string): Promise<Task> {
    return this.api.getTask(taskId);
  }

  async createTask(task: {
    content: string;
    description?: string | null;
    projectId?: string | null;
    sectionId?: string | null;
    parentId?: string | null;
    order?: number | null;
    labels?: string[] | null;
    priority?: number | null;
    assigneeId?: string | null;
    dueString?: string | null;
    dueDate?: string | null;
    dueDatetime?: string | null;
    dueLang?: string | null;
    duration?: number | null;
    durationUnit?: 'minute' | 'day' | null;
  }): Promise<Task> {
    const { dueDatetime, dueDate, content, ...rest } = task;
    const args: any = { content };
    
    // Add non-null optional fields
    Object.entries(rest).forEach(([key, value]) => {
      if (value != null) args[key] = value;
    });
    
    // Handle mutually exclusive due date fields
    if (dueDatetime != null) {
      args.dueDatetime = dueDatetime;
    } else if (dueDate != null) {
      args.dueDate = dueDate;
    }
    
    return this.api.addTask(args);
  }

  async updateTask(taskId: string, updates: {
    content?: string;
    description?: string;
    labels?: string[];
    priority?: number;
    dueString?: string;
    dueDate?: string;
    dueDatetime?: string;
    dueLang?: string;
    assigneeId?: string;
  }): Promise<Task> {
    const { dueDatetime, dueDate, ...rest } = updates;
    const args: any = {};
    
    // Add defined fields (including null for clearing)
    Object.entries(rest).forEach(([key, value]) => {
      if (value !== undefined) args[key] = value;
    });
    
    // Handle mutually exclusive due date fields
    if (dueDatetime !== undefined) {
      args.dueDatetime = dueDatetime;
    } else if (dueDate !== undefined) {
      args.dueDate = dueDate;
    }
    
    return this.api.updateTask(taskId, args);
  }

  async completeTask(taskId: string): Promise<void> {
    await this.api.closeTask(taskId);
  }

  async uncompleteTask(taskId: string): Promise<void> {
    await this.api.reopenTask(taskId);
  }

  async deleteTask(taskId: string): Promise<void> {
    await this.api.deleteTask(taskId);
  }

  async moveTasks(taskIds: string[], destination: {
    projectId?: string;
    sectionId?: string;
    parentId?: string;
  }): Promise<Task[]> {
    const { projectId, sectionId, parentId } = destination;
    const args: any = {};
    
    const provided = [projectId, sectionId, parentId].filter(v => v != null);
    if (provided.length !== 1) {
      throw new Error('Must provide exactly one of projectId, sectionId, or parentId');
    }
    
    if (projectId != null) args.projectId = projectId;
    else if (sectionId != null) args.sectionId = sectionId;
    else if (parentId != null) args.parentId = parentId;
    
    return this.api.moveTasks(taskIds, args);
  }

  // Project methods
  async getProjects(): Promise<Project[]> {
    const response = await this.api.getProjects({
      limit: 200
    });
    return response.results;
  }

  async getProject(projectId: string): Promise<Project> {
    return this.api.getProject(projectId);
  }

  async createProject(project: {
    name: string;
    parentId?: string | null;
    color?: string | null;
    isFavorite?: boolean | null;
    viewStyle?: 'list' | 'board' | null;
  }): Promise<Project> {
    const { name, ...rest } = project;
    const args: any = { name };
    
    // Add non-null optional fields
    Object.entries(rest).forEach(([key, value]) => {
      if (value != null) args[key] = value;
    });
    
    return this.api.addProject(args);
  }

  async updateProject(projectId: string, updates: {
    name?: string;
    color?: string;
    isFavorite?: boolean;
    viewStyle?: 'list' | 'board';
  }): Promise<Project> {
    return this.api.updateProject(projectId, updates);
  }

  async deleteProject(projectId: string): Promise<void> {
    await this.api.deleteProject(projectId);
  }

  // Section methods
  async getSections(projectId?: string): Promise<Section[]> {
    const args: any = {
      limit: 200
    };
    if (projectId) {
      args.projectId = projectId;
    }
    const response = await this.api.getSections(args);
    return response.results;
  }

  async getSection(sectionId: string): Promise<Section> {
    return this.api.getSection(sectionId);
  }

  async createSection(section: {
    name: string;
    projectId: string;
    order?: number | null;
  }): Promise<Section> {
    const { name, projectId, order } = section;
    const args: any = { name, projectId };
    
    if (order != null) args.order = order;
    
    return this.api.addSection(args);
  }

  async updateSection(sectionId: string, updates: {
    name: string;
  }): Promise<Section> {
    return this.api.updateSection(sectionId, updates);
  }

  async deleteSection(sectionId: string): Promise<void> {
    await this.api.deleteSection(sectionId);
  }

  // Label methods
  async getLabels(): Promise<Label[]> {
    const response = await this.api.getLabels({
      limit: 200
    });
    return response.results;
  }

  async getLabel(labelId: string): Promise<Label> {
    return this.api.getLabel(labelId);
  }

  async createLabel(label: {
    name: string;
    color?: string | null;
    order?: number | null;
    isFavorite?: boolean | null;
  }): Promise<Label> {
    const { name, ...rest } = label;
    const args: any = { name };
    
    // Add non-null optional fields
    Object.entries(rest).forEach(([key, value]) => {
      if (value != null) args[key] = value;
    });
    
    return this.api.addLabel(args);
  }

  async updateLabel(labelId: string, updates: {
    name?: string;
    color?: string;
    order?: number;
    isFavorite?: boolean;
  }): Promise<Label> {
    return this.api.updateLabel(labelId, updates);
  }

  async deleteLabel(labelId: string): Promise<void> {
    await this.api.deleteLabel(labelId);
  }

  // Comment methods
  async getComments(params: { taskId?: string; projectId?: string }): Promise<Comment[]> {
    if (params.taskId) {
      const response = await this.api.getComments({ 
        taskId: params.taskId,
        limit: 200 
      });
      return response.results;
    } else if (params.projectId) {
      const response = await this.api.getComments({ 
        projectId: params.projectId,
        limit: 200 
      });
      return response.results;
    } else {
      throw new Error('Either taskId or projectId must be provided');
    }
  }

  async getComment(commentId: string): Promise<Comment> {
    return this.api.getComment(commentId);
  }

  async createComment(comment: {
    content: string;
    taskId?: string | null;
    projectId?: string | null;
    attachment?: any | null;
  }): Promise<Comment> {
    if (!comment.taskId && !comment.projectId) {
      throw new Error('Either taskId or projectId must be provided');
    }
    
    const { content, taskId, projectId, attachment } = comment;
    const args: any = { content };
    
    if (taskId != null) args.taskId = taskId;
    if (projectId != null) args.projectId = projectId;
    if (attachment != null) args.attachment = attachment;
    
    return this.api.addComment(args);
  }

  async updateComment(commentId: string, content: string): Promise<Comment> {
    return this.api.updateComment(commentId, { content });
  }

  async deleteComment(commentId: string): Promise<void> {
    await this.api.deleteComment(commentId);
  }
}

// Re-export types from the official library for convenience
export type { Task, Project, Section, Label, Comment } from '@doist/todoist-api-typescript';